src/text/line.rs

```diff
@@ -586,35 +555,86 @@ impl WidgetRef for Line<'_> {
         };
-        for span in &line.spans {
-            let span_width = span.width() as u16;
-            let span_area = Rect {
-                x,
-                width: span_width.min(area.right().saturating_sub(x)),
-                ..area
-            };
-            span.render(span_area, buf);
-            x = x.saturating_add(span_width);
-            if x >= area.right() {
-                break;
-            }
+    }
+}
+
+/// Renders all the spans of the line that are after the `span_offset` in the line and that are
+/// visible in the given area (starting rendering at `area.x + buf_x_offset`). `buf_x_offset`
+/// probably should just be used to modify area.x instead of being a parameter, but it's here
+/// because that would have made the above code just a little uglier
+///
+/// Algorithm:
+/// 1. For each span in the line:
+///    1. If the span is completely before the `span_offset`, skip it
+///    2. If the span is partially before the `span_offset`, render the right side of the span
+///    3. If the span is completely after the `span_offset`, render the whole span
+/// 2. Indent `area` to account for the width of the rendered span
+/// 3. Update `span_offset` to 0 after the first span is rendered in full
+/// 4. Repeat
+fn render_spans(spans: &[Span], mut area: Rect, buf: &mut Buffer, mut span_offset: u16) {
+    for span in spans {
+        let span_width = span.width() as u16;
+        // ignore spans that are completely before the offset
+        if span_offset >= span_width {
+            span_offset -= span_width;
+            continue;
+        }
+        if area.is_empty() {
+            break;
+        }
+        if span_offset > 0 {
+            // the first span is only partially visible, so truncate the start and render it
+            // starting at span offset
+            let render_width = span_width - span_offset;
+            let (content, display_width) =
+                span.content.unicode_truncate_start(render_width as usize);
+            let display_width = display_width as u16;
+
+            // if the truncation has truncated an initial grapheme, then we need to start rendering
+            // from a position that takes that into account by indenting the start of the area
+            area = indent(area, render_width - display_width);
+            let right = Span::styled(content, span.style);
+            right.render_ref(area, buf);
+
+            area = indent(area, display_width);
+            span_offset = 0; // ensure that the next span is rendered in full
+        } else {
+            // render the whole span
+            span.render_ref(area, buf);
+            area = indent(area, span_width);
         }
     }
 }
 
+/// indents a given `area`'s x value by the given `x_offset`, returning a new `Rect`
+///
+/// TODO: this should probably be move into a pubic method on `Rect` as it's a common operation
+/// that's used in a few places
+const fn indent(area: Rect, x_offset: u16) -> Rect {
```

Comment:
While this shouldn't be public / stabilized yet for the reasons in the other comment I would still move this to the Rect and add `#[must_use]` to ensure its not accidentally misused.

In the rect file it would be `pub(crate)` but it seems way more sorted than having this somewhere in the line file.

Reply:
done in a28718b19